README
NAME
    JSON::T - transform JSON using JsonT

SYNOPSIS
     my $jsont = slurp('foo/bar.js');
     my $input = slurp('foo/quux.json');
     my $JSONT = JSON::T->new($jsont);
     print $JSONT->transform($input);

DESCRIPTION
    This module implements JsonT, a language for transforming JSON-like
    structures, analogous to XSLT in the XML world.

    JsonT is described at <http://goessner.net/articles/jsont/>. JsonT is a
    profile of Javascript; so JsonT needs a Javascript engine to actually
    work. This module provides the engine-neutral stuff, while JSON::T::JE and
    JSON::T::SpiderMonkey provide the necessary glue to hook it up to a couple
    of Javascript engines.

    JSON::T::JE uses the pure Perl Javascript implementation JE.

    JSON::T::SpiderMonkey uses JavaScript::SpiderMonkey which in turn is
    backed by Mozilla's libjs C library.

    This module tries to provide a similar API to XML::Saxon::XSLT2.

  Constructor
    `new($code, $name)`
        Constructs a new JSON::T transformation. $code is the JsonT Javascript
        code. As a JsonT file can contain multiple (potentially unrelated)
        transformations, the name of the particular transformation you want to
        use should also be provided. If $name is omitted, then the name
        "_main" is assumed.

        If you wish to use a particular Javascript implementation, you can
        use, for example:

          JSON::T::SpiderMonkey->new($code, $name)

        Otherwise

          JSON::T->new($code, $name)

        will try to pick a working implementation for you.

  Methods
    `parameters(param1=>$arg1, param2=>$arg2, ...)`
        Sets global variables available to the Javascript code. All arguments
        are treated as strings.

    `transform($input)`
        Run the transformation. The input may be a JSON string, a
        JSON::JOM::Node or a native Perl nested arrayref/hashref structure, in
        which case it will be stringified using the JSON module's to_json
        function. The output (return value) will be a string.

    `transform_structure($input)`
        Like `transform`, but attempts to parse the output as a JSON string
        and return a native Perl arrayref/hashref structure. This method will
        fail if the output is not a JSON string.

    `DOES($role)`
        Like UNIVERSAL's DOES method, but returns true for:

          JSON::T->DOES('XML::Saxon::XSLT2')

        as an aid for polymorphism.

    The following methods also exist for compatibility with XML::Saxon::XSLT2,
    but are mostly useless:

    `transform_document`
    `messages`
    `media_type`
    `version`
    `doctype_system`
    `doctype_public`
    `encoding`

  Javascript Execution Environment
    JSON::T is a profile of Javascript, so is evaluated in an execution
    environment. As this is not a browser environment, many global objects
    familiar to browser Javascript developers are not available. (For example,
    `window`, `document`, `navigator`, etc.)

    A single global object called "JSON" is provided with methods `stringify`
    and `parse` compatible with the well-known json2.js library
    (<http://www.JSON.org/json2.js>), and a method `transform(obj,jsont)` that
    provides a Javascript JsonT implementation.

    A function `print_to_perl` is provided which prints to Perl's STDOUT
    stream.

SUBCLASSING
    Two subclasses are provided: JSON::T::JE and JSON::T::SpiderMonkey, but if
    you need to hook JSON::T up to another Javascript engine, it is relatively
    simple. Just create a Perl class which is a subclass of JSON::T. This
    subclass must implement two required methods and should implement one
    optional method.

    `init`
        Will be passed a newly created object (let's call it $self). It is
        expected to initialise a Javascript execution context for $self, and
        define two Javascript functions: `return_to_perl` (which acts as a
        shim to `$self->_accept_return_value()`) and `print_to_perl` (which
        acts as a shim to `print`). It must then call `SUPER::init`.

    `engine_eval`
        Will be passed an object ($self) and a Javascript string. Must
        evaluate the string in the object's Javascript execution context.

    `parameters`
        This one is optional to implement it. If you don't implement it, then
        users will get a warning message if they try to call `parameters` on
        your subclass.

        Will be passed an object ($self) and a hash of parameters, using the
        following format:

          (
            name1  => 'value1',
            name2  => [ type2 => 'value2' ],
            name3  => [ type3 => 'value3', hint => 'hint value' ],
          )

        This should have the effect of setting:

          var name1 = 'value1';
          var name2 = 'value2';
          var name3 = 'value3';

        in the object's Javascript execution context. Parameter types and
        additional hints may be used to set the correct types in Javascript.

    You are unlikely to need to do anything else when subclassing.

    If you wish `JSON::T->new` to know about your subclass, then push its name
    onto @JSON::T::Implementations.

BUGS
    Please report any bugs to <http://rt.cpan.org/>.

SEE ALSO
    Specification: <http://goessner.net/articles/jsont/>.

    Related modules: JSON, JSON::Path, JSON::GRDDL, JSON::Hyper, JSON::Schema.

    JOM version: JSON::JOM, JSON::JOM::Plugins::JsonT.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

    This module is embeds Stefan Goessner's Javascript implementation of JsonT
    (version 0.9) to do the heavy lifting.

COPYRIGHT AND LICENCE
    Copyright 2006 Stefan Goessner.

    Copyright 2008-2011, 2013 Toby Inkster.

    Licensed under the Lesser GPL:
    <http://creativecommons.org/licenses/LGPL/2.1/>.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
JSON-T
======

Created:      2010-07-29
Home page:    <https://metacpan.org/release/JSON-T>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=JSON-T>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.103	2014-09-11

 [ Packaging ]
 - Switch to Dist::Inkt.

0.102	2013-03-20

 - Updated: no common::sense

0.101	2011-12-05

 [ Documentation ]
 - Create a clear division between the implementation-specific and more
   generic parts of the code. Document how to make implementation
   subclasses.

 [ Packaging ]
 - Module::Package::RDF

 [ Other ]
 - Added: Provide a DOES method.
 - Constructor uses Class::Load to load a suitable subclass.
 - Separate out the JE-specific stuff into JSON::T::JE.
 - Updated: use 5.010
 - Updated: use Object::AUTHORITY

0.100	2010-09-02

 - Accept JSON::JOM as input to transformations.

0.090_04	2010-08-15

 - Force stringification of return value of transform method. Pesky
   JE::String values!

0.090_03	2010-08-10

 - Additional debugging info added to tests.

0.090_02	2010-08-02

 - Switched from JavaScript::SpiderMonkey to JE - SpiderMonkey version is
   retained for those that want it.

0.090_01	2010-07-30

 [ Bug Fixes ]
 - Non-JSON-string input for transform method was not being correctly
   converted to JSON.

0.090_00	2010-07-30


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "transform JSON using JsonT",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Moose::Meta::Class::__ANON__::SERIAL::11 version , CPAN::Meta::Converter version 2.140640",
  "keywords"       => ["ECMAScript", "JSON", "JavaScript", "JsonT", "Transform"],
  "license"        => ["lgpl_2_1"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "JSON-T",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        runtime   => {
                                       recommends => { "JavaScript::SpiderMonkey" => 0 },
                                       requires   => { "Class::Load" => 0, "JE" => 0, "JSON" => "2.00", "Object::AUTHORITY" => 0 },
                                     },
                      },
  "provides"       => {
                        "JSON::T" => { file => "lib/JSON/T.pm", version => 0.103 },
                        "JSON::T::JE" => { file => "lib/JSON/T/JE.pm", version => 0.103 },
                        "JSON::T::SpiderMonkey" => { file => "lib/JSON/T/SpiderMonkey.pm", version => 0.103 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=JSON-T" },
                        homepage     => "https://metacpan.org/release/JSON-T",
                        license      => ["http://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-json-t.git",
                                          web  => "https://github.com/tobyink/p5-json-t",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/JSON-T/project",
                      },
  "version"        => 0.103,
  "x_contributors" => ["Stefan Goessner"],
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

